### PR TITLE
GCL - concurrent requests

### PR DESCRIPTION
**Pull request by Agares**
Signed-off-by: Ramona Łuczkiewicz <rluczkiewicz@wayfair.com>

# Pull request

## Description

Adds support for in-flight concurrency / multiple simultaneous requests.

## Related

Not applicable.

## Checklist

* [x] The RFC, if required, has been submitted and approved
* [x] Any user-facing impact of the changes is reflected in docs.tremor.rs
* [x] The code is tested
* [x] Use of unsafe code is reasoned about in a comment
* [x] Update CHANGELOG.md appropriately, recording any changes, bug fixes, or other observable changes in behavior
* [x] The performance impact of the change is measured (see below)

## Performance

None at this time. Pending feedback production use cases who requested the feature.

### CODE REVIEW

File: src/connectors/google.rs [hunk 1]
```diff
@@ -12,18 +12,89 @@
...
 
+use gouth::Token;
 use std::sync::Arc;
 use tonic::metadata::MetadataValue;
 use tonic::service::Interceptor;
 use tonic::{Request, Status};
 
-pub(crate) struct AuthInterceptor {
-    pub token: Box<dyn Fn() -> ::std::result::Result<Arc<String>, Status> + Send>,
+pub trait TokenProvider: Clone {
```
Comment (Licenser) [inline src/connectors/google.rs:21]: let's avoid pubing things if we can :)
```suggestion
pub(crate) trait TokenProvider: Clone {
```

File: src/connectors/impls/gbq/writer/sink.rs [hunk 2]
```diff
@@ -12,7 +12,11 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use crate::connectors::google::AuthInterceptor;
+#[cfg(not(test))]
```
Comment (Licenser) [inline src/connectors/impls/gbq/writer/sink.rs:15]: I think we can avoid all the cfg test in this file if we generalize `GbqSink` and `AuthInterceptor ` over `TokenProvider` and implement a default method, then during normal operations we can use `GbqSink<GouthTokenProvider>` and during tests `GbqSink<TestTokenProvider>`

File: src/connectors/impls/gcl/writer/sink.rs [hunk 3]
```diff
@@ -13,24 +13,36 @@
 // limitations under the License.
 
 use super::meta;
-use crate::connectors::google::AuthInterceptor;
+#[cfg(not(test))]
```
Comment (Licenser) [inline src/connectors/impls/gcl/writer/sink.rs:16]: Same as for the gbq writer

File: src/connectors/impls/gpubsub/consumer.rs [hunk 4]
```diff
@@ -12,7 +12,10 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use crate::connectors::google::AuthInterceptor;
+#[cfg(not(test))]
```
Comment (Licenser) [inline src/connectors/impls/gpubsub/consumer.rs:15]: same :D

File: src/connectors/impls/gpubsub/producer.rs [hunk 5]
```diff
@@ -12,7 +12,12 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use crate::connectors::google::AuthInterceptor;
+#[cfg(not(test))]
```
Comment (Licenser) [inline src/connectors/impls/gpubsub/producer.rs:15]: Same again ;)

**Review by Licenser — COMMENTED** [review]
👍 not much to improve I would only to to get rid of the code that is `cfg`'ed for tests

File: src/connectors/impls/gbq/writer/sink.rs (hunk 2 above)
Comment (Agares) [reply src/connectors/impls/gbq/writer/sink.rs]: I did that without generalizing, just by using a type alias, I think that's a simpler solution, let me know what you think :)

Comment (Licenser) [reply src/connectors/impls/gbq/writer/sink.rs]: Not really to be honest, there are still 10 compile-time conditionals in the code which makes the whole thing quite hard to read.

Comment (Agares) [reply src/connectors/impls/gbq/writer/sink.rs]: Hmm... I only see a few right now, almost all of them making the TestTokenProvider only compile in tests 🤔. Can you point at the specific problematic lines? Maybe I'm missing something

Comment (Licenser) [reply src/connectors/impls/gbq/writer/sink.rs]: Sure,

25, 27, 31, 36, 49, 66 in google.rs

all of those can be avoided, the Test porovider could move into `mod test` entirely.

I'm not a fan of the type alias either, as compile time switches are, when avoidable, always an addition in complexity that makes code harder to reason about, read, and maintain. We are not writing C where we don't have any other options.

Comment (Agares) [reply src/connectors/impls/gbq/writer/sink.rs]: donezo, unless I missed anything, this should be done now :)

**Review by Licenser — APPROVED** [review]
:rocket:!

**Review by Licenser — APPROVED** [review]
🚀

**Comment by darach** [comment]
FYI. Had to fix DSO back to HEAD~15 for CI so changes are mostly @Agares ( on PTO! ) with bits of myself

**Review by darach — DISMISSED** [review]
LGTM 👍

**Review by mfelsche — DISMISSED** [review]

**Review by darach — APPROVED** [review]
LGTM. Flakiness in code coverage in patch mode 😬

**Review by mfelsche — APPROVED** [review]

**Review by mfelsche — APPROVED** [review]

**Review by mfelsche — APPROVED** [review]

**Review by darach — APPROVED** [review]
Go. Go. Gadget. Armpits! 😬

**Review by mfelsche — APPROVED** [review]

**Pull request merged**